frontend/src/services/HTTPHelper.ts

```diff
@@ -0,0 +1,368 @@
+  [key: string]: string | undefined;
+}
+
+type URLQueryValueType = number | string | (number | string)[];
+
+interface URLQuery {
+  [key: string]: URLQueryValueType;
+}
+
+export abstract class HTTPRequest {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
+    public readonly headers: HTTPHeaderCollection = {},
+    public readonly query: URLQuery = {}
+  ) {}
+}
+
+export class HTTPGETRequest extends HTTPRequest {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(headers: HTTPHeaderCollection = {}, query: URLQuery = {}) {
+    super(headers, query);
+  }
+}
+
+export class HTTPPOSTRequest extends HTTPRequest {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    public readonly body: any,
+    headers: HTTPHeaderCollection = {},
+    query: URLQuery = {}
+  ) {
+    super(headers, query);
+  }
+}
+
+export class HTTPResponse<TResult> {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
+    public readonly result: TResult | null,
+    public readonly headers: HTTPHeaderCollection,
+    public readonly status: number,
+    public readonly statusText: string
+  ) {}
+}
+
+function parseResponseHeaders(headers: string): HTTPHeaderCollection {
+  /* Example:
+   * date: Fri, 08 Dec 2017 21:04:30 GMT\r\n
+   * content-encoding: gzip\r\n
+   * x-content-type-options: nosniff\r\n
+   */
+  const headerKVPs = headers.split('\r\n');
+  const result: HTTPHeaderCollection = {};
+  for (let i = 0; i < headerKVPs.length; i += 1) {
+    // Example: date: Fri, 08 Dec 2017 21:04:30 GMT\r\n
+    const kvp = headerKVPs[i];
+    const indexOfColon = kvp.indexOf(':');
+
+    // A colon was not found in the header line.
+    // This is a malformed header line.
+    if (indexOfColon >= 0) {
+      const key = kvp.slice(0, indexOfColon).trim();
+      const value = kvp.slice(indexOfColon + 1).trim();
+
+      if (key.length > 0 && value.length > 0) {
+        result[key] = value;
+      }
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Builds the query service options.
+ * @param options The partial options that very specified externally.
+ * @returns The built options.
+ */
+function buildOptions(options: Partial<HTTPHelperOptions>): HTTPHelperOptions {
+  let { baseUri } = options;
+
+  // As per the spec of HTTPHelperOptions, we use the base URI of the source the frontend
+  // was loaded from, if no base uri was specified in the options.
+  if (!baseUri) {
+    const { location } = window;
```

Comment:
Isn't that just `window.location.origin`?
Example:
`https://github.com/amosproj/amos-ss2021-project2-context-map/pull/101/files#submit-review` -> `https://github.com`